### PR TITLE
feat: adding playwright testing to starter kit

### LINKED ISSUE #53: Setup Playwright testing
## Describe the solution you'd like

As discussed [here](https://github.com/nickytonline/astro-partykit-starter/discussions/41), we are going to be adding Playwright testing to this repo.

TODO items: 

- [ ] basic setup
- [ ] test for joining a room with username
- [ ] test for joining as anonymous user to the default room
- [ ] test for the modal
- [ ] test for the sending messages
- [ ] update to the `README` about testing with playwright

### PR DESCRIPTION
**Pull request by jdwilkin4**
## Description

This PR does the following: 

- [x] installing Playwright to starter kit
- [x] add test for joining a room with username
- [x] add test for joining as anonymous user to the default room
- [x] add test for the modal
- [x] add test for the sending messages
- [x] updates the `README` to mention testing with playwright

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] 📦 Chore
- [ ] ⏩ Revert

## Related Tickets & Documents
closes #53 

## Mobile & Desktop Screenshots/Recordings

Screenshot 2024-02-03 at 12 39 35 PM

## Steps to QA

1. pull down this branch
2. install dependencies 
3. run `npx playwright test` in the project directory
4. run `npx playwright show-report` to see results on `http://localhost:9323`

## Added to documentation?

- [x] 📜 README.md
- [ ] 📜 CONTRIBUTING.md
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by nickytonline** [comment]
I got the test running via the CLI and the UI of Playwright, but even after installing the dependencies, I had to run `npm playwright install` and then things started working.

```bash
❯ npx playwright test     

Running 12 tests using 5 workers
  1) [chromium] › send-messages.spec.ts:3:1 › user sending a message ─────────

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/chromium-1097/chrome-mac/Chromium.app/Contents/MacOS/Chromium
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  2) [firefox] › join-room.spec.ts:3:1 › join room with username and room number 

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/firefox-1438/firefox/Nightly.app/Contents/MacOS/firefox
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  3) [chromium] › modal.spec.ts:3:1 › user leaving a room ────────────────────

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/chromium-1097/chrome-mac/Chromium.app/Contents/MacOS/Chromium
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  4) [chromium] › join-room.spec.ts:13:1 › join room as anonymous user and default room 

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/chromium-1097/chrome-mac/Chromium.app/Contents/MacOS/Chromium
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  5) [chromium] › join-room.spec.ts:3:1 › join room with username and room number 

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/chromium-1097/chrome-mac/Chromium.app/Contents/MacOS/Chromium
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  6) [webkit] › join-room.spec.ts:3:1 › join room with username and room number 

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/webkit-1967/pw_run.sh
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  7) [firefox] › join-room.spec.ts:13:1 › join room as anonymous user and default room 

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/firefox-1438/firefox/Nightly.app/Contents/MacOS/firefox
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  8) [webkit] › join-room.spec.ts:13:1 › join room as anonymous user and default room 

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/webkit-1967/pw_run.sh
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  9) [firefox] › modal.spec.ts:3:1 › user leaving a room ─────────────────────

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/firefox-1438/firefox/Nightly.app/Contents/MacOS/firefox
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  10) [firefox] › send-messages.spec.ts:3:1 › user sending a message ─────────

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/firefox-1438/firefox/Nightly.app/Contents/MacOS/firefox
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  11) [webkit] › send-messages.spec.ts:3:1 › user sending a message ──────────

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/webkit-1967/pw_run.sh
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  12) [webkit] › modal.spec.ts:3:1 › user leaving a room ─────────────────────

    Error: browserType.launch: Executable doesn't exist at /Users/nicktaylor/Library/Caches/ms-playwright/webkit-1967/pw_run.sh
    ╔═════════════════════════════════════════════════════════════════════════╗
    ║ Looks like Playwright Test or Playwright was just installed or updated. ║
    ║ Please run the following command to download new browsers:              ║
    ║                                                                         ║
    ║     npx playwright install                                              ║
    ║                                                                         ║
    ║ <3 Playwright Team                                                      ║
    ╚═════════════════════════════════════════════════════════════════════════╝


  12 failed
    [chromium] › join-room.spec.ts:3:1 › join room with username and room number 
    [chromium] › join-room.spec.ts:13:1 › join room as anonymous user and default room 
    [chromium] › modal.spec.ts:3:1 › user leaving a room ─────────────────────
    [chromium] › send-messages.spec.ts:3:1 › user sending a message ──────────
    [firefox] › join-room.spec.ts:3:1 › join room with username and room number 
    [firefox] › join-room.spec.ts:13:1 › join room as anonymous user and default room 
    [firefox] › modal.spec.ts:3:1 › user leaving a room ──────────────────────
    [firefox] › send-messages.spec.ts:3:1 › user sending a message ───────────
    [webkit] › join-room.spec.ts:3:1 › join room with username and room number 
    [webkit] › join-room.spec.ts:13:1 › join room as anonymous user and default room 
    [webkit] › modal.spec.ts:3:1 › user leaving a room ───────────────────────
    [webkit] › send-messages.spec.ts:3:1 › user sending a message ────────────
```

File: tests/modal.spec.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { test } from "@playwright/test";
```
Comment (nickytonline) [inline tests/modal.spec.ts:1]: naming (blocking): The name of the test is too generic. Consider calling it something like`leave-room.spec.ts` which is more in line with the names of the other tests.

**Review by nickytonline — CHANGES_REQUESTED** [review]
This is really great @jdwilkin4! Thanks for putting up this PR! Just a few change requests.

**Comment by nickytonline** [comment]
This is a great first step to getting end-to-end (E2E) testing in place for this template project. In another pull request, we should look into making this part of the deployment process so that we can fail a build if the E2E tests fail.

Also, to get it set up for testing in the continuous integration/continuous deployment (CI/CD) flow, we'll need to set up an instance of our party kit app for that. I can help with this if you want to work on this together @jdwilkin4.

**Comment by jdwilkin4** [comment]
>  In another pull request, we should look into making this part of the deployment process so that we can fail a build if the E2E tests fail.

Yeah, I think that is a great idea 👍 

> Also, to get it set up for testing in the continuous integration/continuous deployment (CI/CD) flow, we'll need to set up an instance of our party kit app for that. I can help with this if you want to work on this together @jdwilkin4.

I would be down for that 👍

File: README.md [hunk 2]
```diff
@@ -58,7 +58,9 @@ Any static assets, like images, can be placed in the `public/` directory.
 
 ## End-to-End Testing With Playwright
 
-This starter kit uses [Playwright](https://playwright.dev/) for end-to-end testing. You can find the tests in the `tests/` directory. To run all of the tests, you can run `npx playwright test` from the root of the project. After all of the tests are complete, you will see a report of the tests that passed and failed. Run `npx playwright show-report` to see a visual representation of the test results on `http://localhost:9323`.
+This starter kit uses [Playwright](https://playwright.dev/) for end-to-end testing. To install Playwright, run `npm install playwright` in the terminal.
```
Comment (nickytonline) [inline README.md:61]:
```suggestion
This starter kit uses [Playwright](https://playwright.dev/) for end-to-end testing. To install Playwright, run `npx install playwright` in the terminal.
```

**Review by nickytonline — APPROVED** [review]
Thanks for making the changes @jdwilkin4! One small change request, but I'm going to go ahead and approve.

**Pull request closed without merging**